riscv: pmp: Add API to change region permissions at runtime

fsammoura1980 opened the pull request:
Introduce function `z_riscv_pmp_change_permissions` provides a mechanism to modify the Read, Write, and Execute (R/W/X) permissions of an existing PMP region based on its memory attribute index.

This function locates the PMP entry that corresponds to a region defined in the device tree via memory attributes (as managed by `mem_attr_get_regions`) and updates the PMP configuration register (`pmpcfg`) to reflect the new permissions.

It first iterates through all configured PMP entries, decodes their address and size, and attempts to find a full match for the specified region's start address and size. Once the matching PMP entry is found, its R, W, and X bits are updated based on the provided 'perm' mask, and the change is committed to the PMP configuration registers.

This commit implements a new unit test suite to validate the `z_riscv_pmp_change_permissions` function.

fsammoura1980 committed:
riscv: pmp: Add API to change region permissions at runtime

The new function 'z_riscv_pmp_change_permissions' provides a mechanism
to modify the Read, Write, and Execute (R/W/X) permissions of an
existing PMP region based on its memory attribute index.

Signed-off-by: Firas Sammoura <fsammoura@google.com>

fsammoura1980 committed:
tests: riscv: Implement unit tests for change permissions function

This commit implements a new unit test suite to validate the
`z_riscv_pmp_change_permissions` function.

Signed-off-by: Firas Sammoura <fsammoura@google.com>

npitre reviewed (commented):
A few comments:

- Don't use `ZTEST_USER()` here. None of this should ever be usable from
  user (unprivileged) threads.

- You shouldn't use `csr_write()` when a nice abstraction already exists for
  you with extra validation checks. Please consider
  `write_pmp_entries(entry_index, entry_index+1, false, pmp_addr, pmp_cfg, ARRAY_SIZE(pmp_addr))`.

npitre reviewed (commented):
Why the `MSTATUS_MPRV` addition? Looks like a possible cut-and-paste from
elsewhere. Please justify.

fsammoura1980 commented:
> Look for `arch_irq_lock()` usage. You should justify it with a comment by mentioning that, unlike the other functions, this is not invoked from an atomic context.

Thanks! I modified the code and added a comment.

npitre commented:
> Thanks! I modified the code and added a comment.

You forgot to unlock interrupts if `entry_index == -1`.

kartben commented on include/zephyr/arch/riscv/pmp.h:
```diff
@@ -11,6 +11,33 @@
 
 #ifdef CONFIG_RISCV_PMP
 
+#ifdef CONFIG_MEM_ATTR
```
no guard here please - "[Function declarations that are available only when the option is enabled should be provided unconditionally](https://docs.zephyrproject.org/latest/develop/api/design_guidelines.html#:~:text=Function%20declarations%20that%20are%20available%20only%20when%20the%20option%20is%20enabled%20should%20be%20provided%20unconditionally)."
